src/main/kotlin/org/jitsi/jicofo/jibri/JibriRecorder.kt

```diff
@@ -0,0 +1,198 @@
+            ) ||
+            streamIdIsEmpty && iq.recordingMode == RecordingMode.FILE
+        ) {
+            val sessionId = generateSessionId()
+            try {
+                jibriSession = JibriSession(
+                    this,
+                    conference.roomName,
+                    iq.from,
+                    config.pendingTimeout.seconds,
+                    config.numRetries,
+                    connection,
+                    jibriDetector,
+                    false, null, iq.displayName, iq.streamId, iq.youtubeBroadcastId, sessionId, iq.appData,
+                    logger
+                ).apply {
+                    start()
+                    logger.info("Started Jibri session")
+                }
+                JibriIq.createResult(iq, sessionId)
+            } catch (exc: StartException) {
+                jibriSession = null
+                when (exc) {
+                    is AllBusy -> {
+                        logger.info("Failed to start a Jibri session, all Jibris were busy")
+                        error(iq, XMPPError.Condition.resource_constraint, "all Jibris are busy")
+                    }
+                    is NotAvailable -> {
+                        logger.info("Failed to start a Jibri session, no Jibris available")
+                        error(iq, XMPPError.Condition.service_unavailable, "no Jibri instances available")
+                    }
+                    else -> {
+                        logger.warn("Failed to start a Jibri session:${exc.message}", exc)
+                        error(iq, XMPPError.Condition.internal_server_error, exc.message)
+                    }
+                }
+            }
+        } else if (streamIdIsEmpty && iq.recordingMode == RecordingMode.STREAM) {
+            // Bad request - no stream ID
+            error(iq, XMPPError.Condition.bad_request, "Stream ID is empty or undefined")
+        } else {
+            // Bad request - catch all
+            error(iq, XMPPError.Condition.bad_request, "Invalid recording mode and stream ID combination")
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    override fun onSessionStateChanged(
+        jibriSession: JibriSession,
+        newStatus: Status,
+        failureReason: JibriIq.FailureReason?
+    ) {
+        if (this.jibriSession !== jibriSession) {
+            logger.error("onSessionStateChanged for unknown session: $jibriSession")
+            return
+        }
+        publishJibriRecordingStatus(newStatus, failureReason)
+        if (newStatus == Status.OFF) {
+            this.jibriSession = null
+        }
+    }
+
+    /**
+     * Sends the new jibri status as presence in the conference MUC.
+     */
+    private fun publishJibriRecordingStatus(newStatus: Status, failureReason: JibriIq.FailureReason?) {
+        logger.info("Got jibri status $newStatus and failure $failureReason")
+        val jibriSession = jibriSession
```

Comment:
Why are you copying it to a local here?

Reply:
Because the field might change to `null`.